### ISSUE TITLE
 No 'Access-Control-Allow-Origin' header is present on the requested resource

### ISSUE THREAD

**Issue by hr1383** [mkdynamic/omniauth-facebook#148]
I am using omniauth-facebook with AngularJS and CORS is not working correctly . 

My omniauth.rb is 
<code>Rails.application.config.middleware.use OmniAuth::Builder do
  provider :facebook,"xxxxx", "xxxxx",
     :scope => 'email,user_birthday,read_stream', :display => 'popup'
   
end
</code>
Everything works if i use it as rails app and request. But when i try to call <b>'http:\\localhost:3000\users\auth\facebook"</b> via AngularJS i see following error in JS console 

<blockquote>XMLHttpRequest cannot load https://www.facebook.com/dialog/oauth?client_id=xxxx&display=popup…thday%2Cread_stream&state=3352c1924bdbc9277f7b1070c38d67acf79b529f198323cb. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:3000' is therefore not allowed access. </blockquote>

(The url is not displayed completely in console)

I added following line 
 <code>config.middleware.insert_before Warden::Manager, Rack::Cors</code>
but also this didn't work . 

What is the best way or how can i override headers for OmniAuth ?

**Comment by bindzus** [comment]
I had the exact same problem, using AngularJS with a module called satellizer (http://ngmodules.org/modules/satellizer). At first I stranded much earlier before including the 'rack-cors' gem, so thanks for the hint there, after this gem was added I got the same problem as described above.

I had configured the url property for satellizer to 'http://localhost:3000/auth/facebook' as this would be the route used in the browser for login.

However with satellizer the FB.login is used and the url called afterwards with the code returned from the Facebook SDK, so instead the url property must point to 'http://localhost:3000/auth/facebook/callback'.

My route for '/auth/facebook/callback' now looks like this:

`post '/auth/:provider/callback'       => 'sessions#create'`

And from the session controller action create I return the token to the client side and this seems to work.

**Comment by mkdynamic** [comment]
This is not specific to this library, but rather a general CORS issue. Any README update PRs welcome however.

**Comment by camonunez** [comment]
Found it! I was making a GET request, it should be a POST request.
I'm a newbie, sorry if it was obvious for others

**Comment by fkotsian** [comment]
For anyone who struggled as long as I did... 

@bindzus Is correct. FE libs like Satellizer and VueAuthenticate duplicate the work omniauth-facebook does. When using those libs, you don't need to call `/auth/facebook` to get the OAuth popup window and exchange your code for a token because the FE lib is handling that for you.

Instead you just go straight to `sessions#create` as bindzus did or `/auth/facebook/callback` if you're like me and trying to get DeviseTokenAuth to work

tl;dr
/auth/facebook uses omniauth to get a token, so you don't need to hit it
just get the token using the FE and send it to /auth/facebook/callback or whatever creates sessions

**Comment by kinsbrunner** [comment]
Hi guys,

I am struggling with this. Has anyone come up with any solution? I am working with React client on port 3000 and with Rails API on port 3001.

The error I am getting is:
Access to XMLHttpRequest at 'https://www.facebook.com/v2.10/dialog/oauth?client_id=19176674xxxxxxxxx&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fomniauth%2Ffacebook%2Fcallback&response_type=code&scope=public_profile%2C+email&state=dc8eeac01d631c547c668568798be3edd4a4de849d4e58f2' (redirected from 'http://localhost:3000/auth/facebook') from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

I have tried various things which include enabling rack-cors.

Any advice will be very welcome.
Regards.
Alejandro.

**Comment by chadnewbry** [comment]
It's been a long time since I worked on this, but I seem to remember I was doing what @camonunez  mentioned above.

**Comment by kinsbrunner** [comment]
@chadnewbry not sure why you did what @camonunez said as the /auth/facebook is a GET action and not a POST. Just tried to do it and I got a 404 as that route was not found.
Rails server is running on port 3001 while React client is on port 3000. 
![image](https://user-images.githubusercontent.com/16242832/64478594-a96ca700-d180-11e9-9e92-788fa051abba.png)
There are 2 weird things I can point out:
1. the call should happen through API as that is were this is defined. I found strange that the messages refer to port 3000 while I was expecting them to happen on port 3001.
2. if manually pasting localhost:3001/auth/facebook on the browser, I am seeing that Rails is calling things perfectly fine. This makes me think there is some problem on my routing.....

Any comments would be very welcome!
Regards.
Alejandro.
